lib/mutgen.c

```diff
@@ -182,243 +474,228 @@ mutgen_init_allocator(mutgen_t *self, tsk_table_collection_t *tables)
 }
 
 static int MSP_WARN_UNUSED
-mutgen_add_mutation(mutgen_t *self, node_id_t node, double position,
-        const char *ancestral_state, const char *derived_state)
+mutgen_add_site(mutgen_t *self, double position, bool new,
+        const char *ancestral_state, tsk_size_t ancestral_state_length,
+        const char *metadata, tsk_size_t metadata_length,
+        avl_node_t **avl_nodep)
 {
     int ret = 0;
-    tsk_site_t *site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
-    tsk_mutation_t *mutation = tsk_blkalloc_get(&self->allocator, sizeof(*mutation));
-    avl_node_t* avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
+    char *buff;
+    avl_node_t *avl_node;
+    site_t *site;
 
-    if (site == NULL || mutation == NULL || avl_node == NULL) {
+    avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
+    site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
+    if (site == NULL || avl_node == NULL) {
         ret = MSP_ERR_NO_MEMORY;
         goto out;
     }
-    memset(site, 0, sizeof(*site));
-    memset(mutation, 0, sizeof(*mutation));
+
     site->position = position;
-    site->ancestral_state = ancestral_state;
-    site->ancestral_state_length = 1;
-    mutation->derived_state = derived_state;
-    mutation->derived_state_length = 1;
-    mutation->node = node;
-    mutation->parent = TSK_NULL;
-    site->mutations = mutation;
-    site->mutations_length = 1;
+    site->metadata_length = 0;
+    site->mutations = NULL;
+    site->mutations_length = 0;
+    site->new = new;
+
+    /* ancestral state */
+    buff = tsk_blkalloc_get(&self->allocator, ancestral_state_length);
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    memcpy(buff, ancestral_state, ancestral_state_length);
+    site->ancestral_state = buff;
+    site->ancestral_state_length = ancestral_state_length;
+    /* TODO: do we not have to free buff? */
```

Comment:
No, that's the nice thing about using the blk_alloc. It keeps track of everthing that it has allocated.

Reply:
And it knows when we don't need it any more? We're allocating `buff` lots of times here...

Reply:
``buff`` here is just  a pointer that we assign to the ``ancestral_state``, ``derived_state``, etc.

Reply:
The blk_alloc is a one-way allocator. You ask it for some memory, and that memory will be valid until the blk_alloc itself is freed. That means we only have to make sure that the blk_alloc is freed at the end and don't have to worry about the lifetime's of all these tiny little fragments of memory (it'd be a *nightmare* to track them all via malloc/free)

Reply:
ah, i see!